### PR TITLE
introduce Language DSL v2

### PR DESCRIPTION
**Pull request by OmarTawfik**
## About DSL v2

This introduces DSL v2 that describes the grammar in terms of AST types, not EBNF rules:

- The new model types in [types.rs](https://github.com/OmarTawfik-forks/slang/blob/dsl-v2/crates/codegen/language/definition/src/model/types.rs) are much more accurate/restrictive in describing the semantics of different productions, which removes the need to do a lot of validation that we had for earlier grammars.
- The DSL inside a Rust macro invocation in [definition.rs](https://github.com/OmarTawfik-forks/slang/blob/dsl-v2/crates/solidity/inputs/language/src/definition.rs) are automatically validated on every `cargo check`, are formatted by `rustfmt`, and have definition/references/rename IDE support because of the backing types emitted in [emitter.rs](https://github.com/OmarTawfik-forks/slang/blob/dsl-v2/crates/codegen/language/definition/src/compiler/emitter.rs).

## Behind the Scenes

The magic happens because of the new `codegen_language_internal_macros` crate, which reads the model types, and generates implementations for three things:

1. `Spanned` which rewrites all fields in `Spanned<T>` types that preserve the input token spans for validatation.
2. `ParseInputTokens` which uses `syn` crate to parse the tokens into the backing Rust types.
3. `WriteOutputTokens` which serializes the grammar `Spanned` types into a Rust expression that generates the definition using the original types (without `Spanned<T>`), so that they can be used by client crates.

## Next Steps

We unfortunately now have three sources of truth for the language, that we are manually keeping in sync for now:

1. The YAML grammar ([here](https://github.com/OmarTawfik-forks/slang/blob/dsl-v2/crates/solidity/inputs/language/definition/manifest.yml)) is used for produce the HTML spec.
4. The DSL v1 grammar ([here](https://github.com/OmarTawfik-forks/slang/blob/dsl-v2/crates/solidity/inputs/language/src/dsl.rs)) is used for produce the parser.
5. The DSL v2 grammar ([here](https://github.com/OmarTawfik-forks/slang/blob/dsl-v2/crates/solidity/inputs/language/src/definition.rs)) introduced in this PR, and not used in anything yet.

I will start to delete the YAML grammar, and move `codegen_spec` to use DSL v2, while in parallel starting a discussion about removing DSL v1, as it is a bigger chunk of work that requires coordinating with other ongoing parser/AST work.

## Areas of Improvement

Using `serde` to serialize/deserialize was not originally possible, because its data model does not support token spans, which cannot be serialized, recreated, or persisted outside the context of macro invocations. So I moved to use `syn` for parsing for now. It is working well, although it has a few caveats:

1. I needed to implement a parser, along with custom implementations for `Rc`/`IndexMap`/`Box` and other data structures, that `serde` handles by default.
2. The parser is type driven, which means it is strict, and expects fields to be defined exactly in the same order as the backing Rust types. `struct X { a: u8, b: u8 }` has to be declared as `X(a = 1, b = 2)`, not `X(b = 2, a = 1)`.

I think I found a solution/workaround to the `serde` data model limitation, that will let me remove all these extra implementations and replace it with a `serde` deserializer, but I will look into this in a later iteration, since it is not blocking us right now.

Additionally, validation can still be tightened in a few places, but mostly for keeping the DSL lean/readable, not about correctness, so I will also delay this work for later.

### CODE REVIEW

File: crates/solidity/inputs/language/src/definition.rs [hunk 1]
```diff
@@ -0,0 +1,3854 @@
...
+                    ]
+                ),
+                Topic(
+                    title = "Pragma Directives",
+                    lexical_context = Pragma,
+                    items = [
+                        Struct(
+                            name = PragmaDirective,
+                            fields = (
+                                pragma_keyword = Required(Terminal([PragmaKeyword])),
```
Comment (AntonyBlakey) [inline crates/solidity/inputs/language/src/definition.rs:118]: This seems overly verbose - can `Required~ be the default, and can single element arrays be replaced by list the element?

File: crates/solidity/inputs/language/src/definition.rs [hunk 2]
```diff
@@ -0,0 +1,3854 @@
...
+                            ]
+                        ),
+                        Token(
+                            name = YulDecimalLiteral,
+                            definitions = [TokenDefinition(
+                                scanner = TrailingContext(
+                                    scanner = Choice([
+                                        Atom("0"),
+                                        Sequence([
+                                            Range(inclusive_start = '1', exclusive_end = '9'),
```
Comment (AntonyBlakey) [inline crates/solidity/inputs/language/src/definition.rs:3755]: These `exclusive_end`s should all be `inclusive_end`

File: crates/solidity/inputs/language/src/definition.rs [hunk 3]
```diff
@@ -0,0 +1,3854 @@
...
+                        Struct(
+                            name = EventDefinition,
+                            fields = (
+                                event_keyword = Required(Terminal([EventKeyword])),
+                                name = Required(Terminal([Identifier])),
+                                parameters =
+                                    Optional(reference = NonTerminal(EventParametersDeclaration)),
+                                anonymous_keyword =
+                                    Optional(reference = Terminal([AnonymousKeyword])),
+                                semicolon = Required(Terminal([Semicolon]))
```
Comment (AntonyBlakey) [inline crates/solidity/inputs/language/src/definition.rs:2006]: I prefer this to be a `TerminatedBy` construction so that the error recovery can pick it up.

File: crates/solidity/inputs/language/src/definition.rs [hunk 4]
```diff
@@ -0,0 +1,3854 @@
...
+                        Struct(
+                            name = OverrideSpecifier,
+                            fields = (
+                                override_keyword = Required(Terminal([OverrideKeyword])),
+                                open_paren = Required(Terminal([OpenParen])),
+                                overridden = Required(NonTerminal(OverridePaths)),
+                                close_paren = Required(Terminal([CloseParen]))
+                            )
+                        ),
+                        Separated(
```
Comment (AntonyBlakey) [inline crates/solidity/inputs/language/src/definition.rs:1748]: `SeparatedBy`?

File: crates/solidity/inputs/language/src/definition.rs [hunk 5]
```diff
@@ -0,0 +1,3854 @@
...
+                                    FallbackKeyword,
+                                    ReceiveKeyword
+                                ])),
+                                parameters = Required(NonTerminal(ParametersDeclaration)),
+                                attributes = Required(NonTerminal(FunctionAttributes)),
+                                returns = Optional(reference = NonTerminal(ReturnsDeclaration)),
+                                body = Required(NonTerminal(FunctionBody))
+                            )
+                        ),
+                        Struct(
```
Comment (AntonyBlakey) [inline crates/solidity/inputs/language/src/definition.rs:1660]: I prefer this to be an explicit `DelimitedBy` construct

**Review by AntonyBlakey — COMMENTED** [review]
Seems to be a few drive-by decisions taken that aren't related to the purpose of the PR

File: crates/solidity/inputs/language/src/definition.rs (hunk 2 above)
Comment (OmarTawfik) [reply crates/solidity/inputs/language/src/definition.rs:3755]: Good catch. Fixed!

File: crates/solidity/inputs/language/src/definition.rs (hunk 1 above)
Comment (OmarTawfik) [reply crates/solidity/inputs/language/src/definition.rs:118]: Should be able to have these fine controls once we move to serde, using things like `#[serde(flatten)]` and `#[serde(default)]`, but for now, I suggest keeping the hand-written parser simple/straight-forward. It is expensive to add more capabilities for it only to throw it away later.

File: crates/solidity/inputs/language/src/definition.rs (hunk 4 above)
Comment (OmarTawfik) [reply crates/solidity/inputs/language/src/definition.rs:1748]: I suggest using `Separated` since it describes the item being separated (higher importance to describe/annotate). The term `SeparatedBy` refers to the separator itself (lower importance). This is also reflected in the backing rust type `SeparatedItem`, instead of `SeparatedByItem`.

For example in `Parameters`, the focus here is on the `Parameter` being separated, not the `Comma`. This is the thing we would typically extract/use in further passes.

File: crates/solidity/inputs/language/src/definition.rs (hunk 3 above)
Comment (OmarTawfik) [reply crates/solidity/inputs/language/src/definition.rs:2006]: Discussing offline.

File: crates/solidity/inputs/language/src/definition.rs (hunk 5 above)
Comment (OmarTawfik) [reply crates/solidity/inputs/language/src/definition.rs:1660]: Discussing offline.

File: crates/codegen/language/definition/src/compiler/versions.rs [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+                first
+            } else {
+                break;
+            };
+
+            let second = if let Some(second) = second_iter.peek() {
+                second
+            } else {
+                break;
+            };
```
Comment (Xanewok) [inline crates/codegen/language/definition/src/compiler/versions.rs:115]: Small suggestion (and below): 

```suggestion
            let (first, second) = match (first_iter.peek(), second_iter.peek()) {
                (Some(first), Some(second)) => (first, second),
                _ => break,
            };
```

File: crates/codegen/language/definition/src/compiler/versions.rs [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        }
+
+        return instance;
+    }
+
+    pub fn is_empty(&self) -> bool {
+        return self.ranges.is_empty();
+    }
+
+    pub fn add(&mut self, range: &VersionRange) {
```
Comment (Xanewok) [inline crates/codegen/language/definition/src/compiler/versions.rs:95]: Is there a reason why this is mutable but the difference is immutable but creates a new value?

File: crates/codegen/language/definition/src/compiler/versions.rs [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+        if range.is_empty() {
+            return;
+        }
+
+        let mut first_iter = self.ranges.iter().peekable();
+        let mut second_iter = once(range).into_iter().peekable();
+        let mut ranges = vec![];
+
+        loop {
```
Comment (Xanewok) [inline crates/codegen/language/definition/src/compiler/versions.rs:104]: It took me a while to understand how we're managing the two streams and conditionally consume the elements from either or both. 

I'm not sure what the exact semantics should be when it comes to merging disjoint ranges or normalizing the representation. Since this is "pure" and very mechanical code, it'd be great to have some unit tests to serve as an example.

For instance, I expected the following result to be an empty set but it seems not to be:
```rust
let mut example = VersionSet::empty();
example.add(&range!(0,8,0 till 0,9,0));
example.add(&range!(1,0,0 till 1,1,0));

example.add(&range!(0,9,0 till 1,0,0));
dbg!(&example.difference(&VersionSet::single(range!(0,8,0 till 1,1,0))));
// &example.difference(&VersionSet::single(range!(0, 8, 0 till 1, 1, 0))) = VersionSet {
//     ranges: [
//         VersionRange {
//             inclusive_start: Version {
//                 major: 1,
//                 minor: 0,
//                 patch: 0,
//             },
//             exclusive_end: Version {
//                 major: 1,
//                 minor: 1,
//                 patch: 0,
//             },
//         },
//     ],
// }
```
Maybe simplifying the zipping/state management would help catch that?

File: crates/codegen/language/definition/src/model/identifier.rs [hunk 9]
```diff
@@ -0,0 +1,45 @@
+use serde::Serialize;
+use std::ops::Deref;
+
+/// A wrapper type to make sure the DSL token is written as an identifier instead of a string literal.
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct Identifier {
```
Comment (Xanewok) [inline crates/codegen/language/definition/src/model/identifier.rs:6]: I'm going to risk a dumb question - could we use proc_macro2/syn::Ident here directly, since we only seem to use it with the macro and we almost always want it with a span?

**Review by Xanewok — COMMENTED** [review]
Looks good! Left some nitpicks.

I'm slightly nervous that we're de facto writing compiler/analysis for the compiler introducing even more machinery for the language definition itself (despite knowing it's a good decision). I think that our immediate priority right now should be to streamline this ASAP as this hurts code navigation (which language definition framework is this file part of, again?) and increases the required mental capacity to move around and work inside the code.

File: crates/codegen/language/definition/src/compiler/analysis/definitions.rs [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+impl Analysis {
+    pub fn analyze_definitions(&mut self) {
+        self.collect_top_level_items();
+
+        self.check_enum_items();
+        self.check_precedence_items();
+    }
+
+    fn collect_top_level_items(&mut self) {
+        for item in self.language.clone().items() {
```
[the remark this answers is not in the document]
Comment (OmarTawfik) [reply crates/codegen/language/definition/src/compiler/analysis/definitions.rs:22]: Unfortunately it is not, as `analysis` is borrowed as a mutable otherwise.

File: crates/codegen/language/definition/src/compiler/analysis/definitions.rs [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+            let name = match item {
```
[the remark this answers is not in the document]
Comment (OmarTawfik) [reply crates/codegen/language/definition/src/compiler/analysis/definitions.rs:23]: Sure.

File: crates/codegen/language/definition/src/compiler/analysis/references.rs [hunk 12]
```diff
@@ -0,0 +1,503 @@
+use crate::{
```
[the remark this answers is not in the document]
Comment (OmarTawfik) [reply crates/codegen/language/definition/src/compiler/analysis/references.rs:1]: Sure. Happy to do that in the next iteration.

File: crates/codegen/language/definition/src/compiler/analysis/references.rs [hunk 13]
```diff
@@ -0,0 +1,503 @@
...
+
+        if let (Some(first), Some(second)) = (first, second) {
+            if first >= second {
+                self.errors
+                    .add(first, &Errors::UnorderedVersionPair(first, second));
+            }
+        }
+    }
+
+    fn check_reference(
```
[the remark this answers is not in the document]
Comment (OmarTawfik) [reply crates/codegen/language/definition/src/compiler/analysis/references.rs:401]: Sounds good.

File: crates/codegen/language/definition/src/compiler/versions.rs (hunk 7 above)
Comment (OmarTawfik) [reply crates/codegen/language/definition/src/compiler/versions.rs:95]: Missed that from an earlier refactoring. Good catch. Will fix it here.

Comment (OmarTawfik) [reply crates/codegen/language/definition/src/compiler/versions.rs:95]: Not always. Otherwise, this serves as an accumulator.

File: crates/codegen/language/definition/src/compiler/versions.rs (hunk 6 above)
Comment (OmarTawfik) [reply crates/codegen/language/definition/src/compiler/versions.rs:115]: Sounds good.

File: crates/codegen/language/definition/src/compiler/versions.rs (hunk 7 above)
Comment (OmarTawfik) [reply crates/codegen/language/definition/src/compiler/versions.rs:95]: Left from an earlier refactoring. Good catch. Fixed.

Comment (OmarTawfik) [reply crates/codegen/language/definition/src/compiler/versions.rs:95]: Not always. It sometimes serves as an accumulator, instead of cloning `self` when it is not changed.

File: crates/codegen/language/definition/src/internals/errors/mod.rs [hunk 14]
```diff
@@ -0,0 +1,85 @@
+use proc_macro2::{Span, TokenStream};
```
[the remark this answers is not in the document]
Comment (OmarTawfik) [reply crates/codegen/language/definition/src/internals/errors/mod.rs:1]: Sure.

File: crates/codegen/language/definition/src/internals/spanned/mod.rs [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+
+impl<T: std::cmp::Eq> std::cmp::Eq for Spanned<T> {}
+
+impl<T: std::hash::Hash> std::hash::Hash for Spanned<T> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        return self.value.hash(state);
+    }
+}
+
+impl<T: std::cmp::Ord> std::cmp::Ord for Spanned<T> {
```
[the remark this answers is not in the document]
Comment (OmarTawfik) [reply crates/codegen/language/definition/src/internals/spanned/mod.rs:45]: Sounds good.

File: crates/codegen/language/definition/src/model/identifier.rs (hunk 9 above)
Comment (OmarTawfik) [reply crates/codegen/language/definition/src/model/identifier.rs:6]: We cannot use it instead of `Identifier`, as we cannot implement other traits for this external type.
We cannot wrap it inside `Identifier` either, as we would have to special case it from `internal_macros` in order for it to not be wrapped unnecissarily in `Spanned<>` again.
Also, it is not `serde` serializable, at least not without workarounds, which we need for the entire model.

File: crates/solidity/inputs/language/src/definition.rs (hunk 3 above)
Comment (OmarTawfik) [reply crates/solidity/inputs/language/src/definition.rs:2009]: Added `error_recovery` field.

File: crates/solidity/inputs/language/src/definition.rs (hunk 5 above)
Comment (OmarTawfik) [reply crates/solidity/inputs/language/src/definition.rs:1663]: Added `error_recovery` field.

File: crates/codegen/language/definition/src/compiler/analysis/definitions.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Xanewok) [reply crates/codegen/language/definition/src/compiler/analysis/definitions.rs]: Hm, it was redundant in f11542a8de6c6b3d3b586ffe4a60e25d76d603df and now we can remove it with

```diff
diff --git a/crates/codegen/language/definition/src/compiler/analysis/definitions.rs b/crates/codegen/language/definition/src/compiler/analysis/definitions.rs
index 430d8587..6c30e6fc 100644
--- a/crates/codegen/language/definition/src/compiler/analysis/definitions.rs
+++ b/crates/codegen/language/definition/src/compiler/analysis/definitions.rs
@@ -18,7 +18,7 @@ pub fn analyze_definitions(analysis: &mut Analysis) {
 }
 
 fn collect_top_level_items(analysis: &mut Analysis) {
-    for item in analysis.language.clone().items() {
+    for item in analysis.language.items() {
         let name = get_item_name(item);
         let defined_in = calculate_defined_in(analysis, item);
 
@@ -111,7 +111,7 @@ fn get_item_name(item: &Item) -> &Spanned<Identifier> {
     }
 }
 
-fn calculate_defined_in(analysis: &mut Analysis, item: &Item) -> VersionSet {
+fn calculate_defined_in(analysis: &Analysis, item: &Item) -> VersionSet {
     return match item {
         Item::Struct { item } => VersionSet::from_range(calculate_enablement(
             analysis,
@@ -157,7 +157,7 @@ fn calculate_defined_in(analysis: &mut Analysis, item: &Item) -> VersionSet {
 }
 
 fn calculate_enablement(
-    analysis: &mut Analysis,
+    analysis: &Analysis,
     enabled_in: &Option<Spanned<Version>>,
     disabled_in: &Option<Spanned<Version>>,
 ) -> VersionRange {

```

but it's a miniscule thing

File: crates/codegen/language/definition/src/model/identifier.rs (hunk 9 above)
Comment (Xanewok) [reply crates/codegen/language/definition/src/model/identifier.rs:6]: Got it, thanks! Maybe it'd be worthwhile to add a comment why we're not using `proc_macro2::Ident` directly?

File: crates/codegen/language/definition/src/compiler/versions.rs (hunk 8 above)
Comment (Xanewok) [reply crates/codegen/language/definition/src/compiler/versions.rs:106]: So is this expected? Do you want to leave that for a follow-up?

Comment (OmarTawfik) [reply crates/codegen/language/definition/src/compiler/versions.rs:106]: Good catch. Fixed the corner case and added tests. Bug existed since the legacy implementation in `crates/codegen/schema/src/validation/visitors/version_set.rs`, so I'm fixing that as well.

Luckily, it didn't miss anything, and the grammar is still correct.

Comment (OmarTawfik) [reply crates/codegen/language/definition/src/compiler/versions.rs:106]: Good catch. Fixed the corner case and added tests.

Luckily, it didn't miss anything, and the grammar is still correct.

**Review by Xanewok — APPROVED** [review]
Looks good, thanks a lot for the work on the next version of the DSL! Let's merge this once the CI is green 🎉
